src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2221 @@
+    if (!(options->HSLayer2Nodes || options->HSLayer3Nodes))
+      return 0;
+  }
+
+  if (!(circpad_circuit_state(circ) & machine->conditions.state_mask))
+    return 0;
+
+  if (circuit_get_cpath_opened_len(circ) < machine->conditions.min_hops)
+    return 0;
+
+  return 1;
+}
+
+/**
+ * Returns a minimized representation of the circuit state.
+ *
+ * The padding code only cares if the circuit is building,
+ * opened, or used for streams. This returns which one is true.
+ */
+static inline
+circpad_circuit_state_t
+circpad_circuit_state(origin_circuit_t *circ)
+{
+  if (circ->p_streams)
+    return CIRCPAD_CIRC_STREAMS;
+
+  /* We use has_opened to prevent cannibialized circs from flapping. */
+  if (circ->has_opened)
+    return CIRCPAD_CIRC_OPENED;
+
+  return CIRCPAD_CIRC_BUILDING;
+}
+
+/**
+ * Convert a normal circuit purpose into a bitmask that we can
+ * use for determining matching circuits.
+ */
+static inline
+circpad_purpose_mask_t
+circpad_circ_purpose_to_mask(uint8_t circ_purpose)
+{
+  /* Treat OR circ purposes as ignored. They should not be passed here*/
+  if (BUG(circ_purpose <= CIRCUIT_PURPOSE_OR_MAX_)) {
+    return 0;
+  }
+
+  /* Treat new client circuit purposes as "OMG ITS EVERYTHING".
+   * This also should not happen */
+  if (BUG(circ_purpose - CIRCUIT_PURPOSE_OR_MAX_ - 1 > 32)) {
+    return CIRCPAD_PURPOSE_ALL;
+  }
+
+  /* Convert the purpose to a bit position */
+  return 1 << (circ_purpose - CIRCUIT_PURPOSE_OR_MAX_ - 1);
+}
+
+/**
+ * Shut down any machines whose conditions no longer match
+ * the current circuit.
+ */
+static void
+circpad_shutdown_old_machines(origin_circuit_t *on_circ)
+{
+  circuit_t *circ = TO_CIRCUIT(on_circ);
+
+  for (int i = 0; i < CIRCPAD_MAX_MACHINES; i++) {
+    if (!circ->padding_machine[i])
+      continue;
+
+    if (!circpad_machine_conditions_met(on_circ,
+                                        circ->padding_machine[i])) {
+      // Clear machineinfo (frees timers)
+      circpad_circuit_machineinfo_free_idx(circ, i);
+      // Send padding negotiate stop
+      circpad_negotiate_padding(on_circ,
+                                circ->padding_machine[i]->machine_num,
+                                circ->padding_machine[i]->target_hopnum,
+                                CIRCPAD_COMMAND_STOP);
+    }
+  }
+}
+
+/**
+ * Negotiate new machines that would apply to this circuit.
+ */
+static void
+circpad_add_matching_machines(origin_circuit_t *on_circ)
+{
+  circuit_t *circ = TO_CIRCUIT(on_circ);
+
+#ifdef TOR_UNIT_TESTS
+  /* Tests don't have to init our padding machines */
+  if (!origin_padding_machines)
+    return;
+#endif
+
+  for (int i = 0; i < CIRCPAD_MAX_MACHINES; i++) {
+    if (circ->padding_info[i])
+      continue;
+
+    /* We have a free index */
+    SMARTLIST_FOREACH_REVERSE_BEGIN(origin_padding_machines,
+                                    circpad_machine_t *,
+                                    machine) {
+      if (machine->machine_index == i &&
```

Comment:
Relevant to the comment above, what's the index check here? Can we replace that with a pointer to the global machine or a function?

Reply:
Commented in eb11886fccada2b56d87cf85fda5fb553298bfbf -- machine indexes need to be specified explicitly so that when there are two machines on the circuit, events arrive at them in a predictable deterministic order that they are designed to expect -- because they specify which index they go into.

Reply:
Thanks for the comments.

Is it also possible to document `origin_padding_machines` and `relay_padding_machines` and also specify where those come from and how they get initialized?

Reply:
1208109dc4a20b838c11c827d1a24abfdf253e49